### PR TITLE
Plugin Import Prototype

### LINKED ISSUE #5: Import plugins
From sandboxes created by [ianjennings](https://github.com/ianjennings): babel/sandboxes#13  

Should be able to import plugin into the codesandbox from the babel source / source code repo.

### PR DESCRIPTION
**Pull request by garrettluu**
Addresses #5 

Made a prototype for including and excluding default plugins in a similar manner to the Babel REPL. This will be kinda long but I want to explain some of the design decisions that I made that might have impacts throughout the project! Here are some important changes to note:

1. We are using our own JSON format for all internal purposes so that we can easily display data in the front end. The default list of plugins looks like this with the new format:
```js
export const plugins = {
         "@babel/plugin-external-helpers": {
           name: "@babel/plugin-external-helpers",
           ref: require("@babel/plugin-external-helpers"), // will be deleted 
           description: "does this",
           fileLocation: "/static/plugins/@babel/plugin-external-helpers.js",
           defaultConfig: {
             method: "usage-global",
             targets: {
               edge: 16,
             },
           },
         },
         "babel-plugin-polyfill-corejs2": {
           name: "babel-plugin-polyfill-corejs2",
           ref: require("babel-plugin-polyfill-corejs2"),
           description: "does this",
           fileLocation: "/static/plugins/babel-plugin-polyfill-corejs2.js",
           defaultConfig: {
             method: "usage-global",
             targets: {
               edge: 16,
             },
           },
         },
//...
}
```
and the plugin config will have this format:
```js
const CONFIG = [
  {
    plugins: [
      {
         name: "babel-plugin-polyfill-corejs3",
         description: "does this",
         fileLocation: "babel-plugin-polyfill-corejs3",
         defaultConfig: {
           method: "usage-global",
           targets: {
             edge: 16,
           }
         },
       },
    ],
  },
  // {},
];
```
I also wrote functions `convertToBabelConfig` and `convertToJsonConfig` to convert between our own format and the official Babel format. **We will only convert to Babel format when interfacing with Babel or exporting the config.** Otherwise, use JSON form for everything else for consistency!

2. Plugin imports for default plugins are done by looking at the `fileLocation` property and appending the script directly to the page. I'm not too sure about this one, since we can technically just directly embed a `script` tag into `index.html` and it will have the same effect (we would also remove the `fileLocation` property since that won't be used anymore). Also, I don't really see an advantage to doing over using the good ol' `require`.
```js
function importDefaultPlugins() {
  Object.keys(plugins).forEach((pluginName) => {
    const script = document.createElement("script");
    script.src = plugins[pluginName].fileLocation;
    script.async = false;
    document.head.appendChild(script);
  });
  console.log(window);
}
```
Babel standalone will then register the plugin so we can use it.
```js
function registerDefaultPlugins() {
  Babel.registerPlugin(
    "babel-plugin-polyfill-corejs3",
    window.babelPluginPolyfillCorejs3
  );
  Babel.registerPlugin(
    "babel-plugin-polyfill-corejs2",
    window.babelPluginPolyfillCorejs2
  );
  Babel.registerPlugin(
    "@babel/plugin-external-helpers",
    window._babel_pluginExternalHelpers
  );
  Babel.registerPlugin(
    "babel-plugin-polyfill-es-shims",
    window.babelPluginPolyfillEsShims
  );
  Babel.registerPlugin(
    "babel-plugin-polyfill-regenerator",
    window.babelPluginPolyfillRegenerator
  );
}
```
Try it out here: https://doo40.csb.app/

Features that still need to be implemented:

- ~[ ] Dynamic plugin imports of other plugins not in the default list, using bundle.run~
- ~[ ] Allow editing of the config file without breaking everything~
- [x] Do the same thing, but for presets too

### CODE REVIEW

**Comment by ianjennings** [comment]
This is looking good, it looks like you almost have the full feature ready! We can merge a prototype and fix it later, so no worries if it's not perfect.

Please merge `master` in soon. I can help you on Monday.

Let me know when you want a full review.

**Comment by ianjennings** [comment]
> Plugin imports for default plugins are done by looking at the fileLocation property and appending the script directly to the page. I'm not too sure about this one, since we can technically just directly embed a script tag into index.html and it will have the same effect (we would also remove the fileLocation property since that won't be used anymore). Also, I don't really see an advantage to doing over using the good ol' require.

`require` means that the page needs to be compiled with the plugin. If require all the plugins by default, it will make the package size bigger.

File: public/index.html [hunk 1]
```diff
@@ -1,5 +1,10 @@
 <!DOCTYPE html>
 <html lang="en">
+  <head>
+    <meta charset="UTF-8" />
+    <meta name="viewport" content="width=device-width, initial-scale=1.0" />
+    <meta http-equiv="X-UA-Compatible" content="ie=edge" />
+    <title>Babel Test Playground</title>
```
Comment (ianjennings) [inline public/index.html:7]: 🤔 wonder how this happened, I could swear the title is already set

File: src/components/App.js [hunk 2]
```diff
@@ -11,17 +11,76 @@ import { useDebounce } from "../utils/useDebounce";
...
+/**
+ * Converts internal json plugin/preset config to babel form
+ * @param {Object} jsonConfig 
+ */
+export function convertToBabelConfig(jsonConfig) {
+  let result = {plugins: [], presets: []};
+  result.plugins = jsonConfig.plugins?.map(plugin => [plugin.name, plugin.defaultConfig]);
+  result.presets = jsonConfig.presets?.map(preset => [preset.name, preset.defaultConfig]);
+  return result;
+}
```
Comment (ianjennings) [inline src/components/App.js:27]: 👍

File: src/components/App.js [hunk 3]
```diff
@@ -11,17 +11,76 @@ import { useDebounce } from "../utils/useDebounce";
...
+  );
+  Babel.registerPlugin(
+    "babel-plugin-polyfill-es-shims",
+    window.babelPluginPolyfillEsShims
+  );
+  Babel.registerPlugin(
+    "babel-plugin-polyfill-regenerator",
+    window.babelPluginPolyfillRegenerator
+  );
+}
```
Comment (ianjennings) [inline src/components/App.js:72]: 👍

File: src/components/CompiledOutput.js [hunk 4]
```diff
@@ -1,10 +1,18 @@
...
+import * as Babel from "@babel/standalone";
 import { processOptions } from "../standalone";
 import { gzipSize } from "../gzip";
 import { Wrapper, Code, Config } from "./styles";
 import { useDebounce } from "../utils/useDebounce";
 
+import {
+  convertToBabelConfig,
+  //   importDefaultPlugins,
+  //   registerDefaultPlugins
```
Comment (ianjennings) [inline src/components/CompiledOutput.js:11]: Either remove these comments or add more details explaining what they're doing here.

File: src/components/CompiledOutput.js [hunk 5]
```diff
@@ -35,7 +46,75 @@ export function CompiledOutput({
...
+            <input
+              name={pluginName}
+              type="checkbox"
+              onChange={handlePluginChange}
+            />
+            <label>{plugin.name}</label>
+        </div>
+      );
+    });
+  }
```
Comment (ianjennings) [inline src/components/CompiledOutput.js:65]: can't wait to style this

File: src/index.js [hunk 6]
```diff
@@ -18,19 +18,38 @@ const SOURCE = `Promise.allSettled([p1, p2]).finally(() => {
...
+      //   name: "babel-plugin-polyfill-corejs3",
+      //   description: "does this",
+      //   fileLocation: "babel-plugin-polyfill-corejs3",
+      //   defaultConfig: {
+      //     method: "usage-global",
+      //     targets: {
+      //       edge: 16,
+      //     }
+      //   },
+      // },
```
Comment (ianjennings) [inline src/index.js:40]: Remove comments or describe what this is. Looks like this is trying to show what these values will be when the function runs.

File: src/standalone.js [hunk 7]
```diff
@@ -1,18 +1,15 @@
...
 
-export function transpilePlugin(pluginString) {
+export function transpilePlugin(pluginString, presets) {
   return Babel.transform(pluginString, {
     babelrc: false,
     configFile: false,
     ast: false,
     highlightCode: false,
-    presets: [availablePresets["@babel/preset-env"]],
+    presets: presets
```
Comment (ianjennings) [inline src/standalone.js:12]: When the property name and variable are the same, you can just supply `presets` without the value. Being explicit is good though.

File: src/standalone.js [hunk 8]
```diff
@@ -22,7 +19,7 @@ export default function compileModule(code, globals = {}) {
   let module = { exports };
   let globalNames = Object.keys(globals);
   let keys = ["module", "exports", ...globalNames];
-  let values = [module, exports, ...globalNames.map(key => globals[key])];
+  let values = [module, exports, ...globalNames.map((key) => globals[key])];
```
Comment (ianjennings) [inline src/standalone.js:22]: I'm guessing this is a linter thing.

**Review by ianjennings — APPROVED** [review]
LGTM!

File: src/components/CompiledOutput.js (hunk 4 above)
Comment (garrettluu) [reply src/components/CompiledOutput.js:11]: Oops, there's some commented snippets of code here and there that I forgot to remove.

**Pull request merged**